### PR TITLE
Feature/allow users to see course total score

### PR DESCRIPTION
**Pull request by krusche**

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/ParticipationRepository.java [hunk 1]
```diff
@@ -16,11 +16,14 @@
...
-
     @Query("select p from Participation p where p.exercise.course.id = :courseId")
     List<Participation> findByCourseId(@Param("courseId") Long courseId);
 
+    @Query("select participation from Participation participation where participation.exercise.course.id = :courseId and participation.student.login= :username")
+    List<Participation> findByCourseIdAndStudentLogin(@Param("courseId") Long courseId, @Param("username") String username);
+
+    List<Participation> findByExerciseId(@Param("exerciseId") Long exerciseId);
+
     Participation findOneByExerciseIdAndStudentLogin(Long exerciseId, String username);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ParticipationRepository.java]: Can you revert these changes as well, if they are not needed?

File: src/main/java/de/tum/in/www1/artemis/service/CourseService.java [hunk 2]
```diff
@@ -22,16 +23,21 @@
...
     public CourseService(CourseRepository courseRepository,
                          UserService userService,
                          ExerciseService exerciseService,
-                         AuthorizationCheckService authCheckService) {
+                         AuthorizationCheckService authCheckService, ParticipationRepository participationRepository) {
         this.courseRepository = courseRepository;
+        this.participationRepository = participationRepository;
         this.userService = userService;
         this.exerciseService = exerciseService;
         this.authCheckService = authCheckService;
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/CourseService.java]: Can you revert these changes as well, if they are not needed?

File: src/main/webapp/app/entities/course/course-score-calculation-component.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+    ngOnInit() {
+        this.subscription = this.route.params.subscribe(params => {
+            this.courseId = params['id'];
+        });
+
+        this.course = this.courseCalculationService.getCourse(this.courseId);
+
+        if(this.course == undefined) {
+            console.log('Blank');
```
Comment (krusche) [inline src/main/webapp/app/entities/course/course-score-calculation-component.ts]: please remove log statements for the production use

File: src/main/java/de/tum/in/www1/artemis/repository/ParticipationRepository.java (hunk 1 above)
Comment (LaKorita) [reply src/main/java/de/tum/in/www1/artemis/repository/ParticipationRepository.java]: Done :)

File: src/main/java/de/tum/in/www1/artemis/service/CourseService.java (hunk 2 above)
Comment (LaKorita) [reply src/main/java/de/tum/in/www1/artemis/service/CourseService.java]: Done :)

File: src/main/webapp/app/entities/course/course-score-calculation-component.ts (hunk 3 above)
Comment (LaKorita) [reply src/main/webapp/app/entities/course/course-score-calculation-component.ts]: My bad. Removed :D

File: src/main/java/de/tum/in/www1/artemis/service/CourseService.java [hunk 4]
```diff
@@ -22,6 +23,9 @@
 
     private final Logger log = LoggerFactory.getLogger(CourseService.class);
 
+    private static final double INITIAL_TOTAL_SCORE = 0.0;
+    private static final double SCORE_NORMALIZATION_VALUE = 0.01;
+
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/service/CourseService.java]: can you also revert these changes please?

File: src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java [hunk 5]
```diff
@@ -163,7 +163,6 @@ public JsonNode getAllCoursesForDashboard(Principal principal) {
             courseJson.set("exercises", exercisesJson);
             coursesJson.add(courseJson);
         }
-
         // return json array of courses
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/CourseResource.java]: can you also revert these changes please?

File: src/main/java/de/tum/in/www1/artemis/repository/ParticipationRepository.java [hunk 6]
```diff
@@ -16,11 +16,11 @@
 @Repository
 public interface ParticipationRepository extends JpaRepository<Participation, Long> {
 
-    List<Participation> findByExerciseId(@Param("exerciseId") Long exerciseId);
-
     @Query("select p from Participation p where p.exercise.course.id = :courseId")
     List<Participation> findByCourseId(@Param("courseId") Long courseId);
 
+    List<Participation> findByExerciseId(@Param("exerciseId") Long exerciseId);
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/repository/ParticipationRepository.java]: can you also revert these changes please?

File: src/main/webapp/app/entities/course/course.service.ts [hunk 7]
```diff
@@ -310,6 +311,158 @@ export class CourseScoresService {
     }
 }
 
+@Injectable()
+export class CourseScoreCalculationService {
+
```
Comment (krusche) [inline src/main/webapp/app/entities/course/course.service.ts]: Could we put this service into its own file?

File: src/main/webapp/app/entities/course/course.service.ts [hunk 8]
```diff
@@ -320,32 +320,50 @@ export class CourseScoreCalculationService {
...
-            }
-        }
-        scores.set('absoluteScore', Math.round(absoluteScore * 100) / 100);
+        courseExercises.forEach( exercise => {
+          if (exercise.maxScore !== null) {
+              maxScore = maxScore + exercise.maxScore;
+              const participation: Participation = this.getParticipationForExercise(exercise);
+              if (participation !== undefined) {
+                  const result: Result = this.getResultForParticipation(participation, exercise.dueDate);
+                  absoluteScore = absoluteScore + result.score;
```
Comment (LaKorita) [inline src/main/webapp/app/entities/course/course.service.ts]: Could you please revert the formula again? The absolute score now is wrongly calculated...

Comment (LaKorita) [inline src/main/webapp/app/entities/course/course.service.ts]: Could you please revert the changes you did in the score calculation formulas? They are wrongly calculated now, even the relative score...Just tested it locally :/ @Balgorn

Comment (LaKorita) [reply src/main/webapp/app/entities/course/course.service.ts]: `absoluteScore + result.score * this.SCORE_NORMALIZATION_VALUE * exercise.maxScore;`

This should be how we calculate the absolute score

File: src/main/webapp/app/entities/course/course.service.ts [hunk 9]
```diff
@@ -310,6 +311,176 @@ export class CourseScoresService {
...
+        if (maxScore > 0) {
+            scores.set('relativeScore', this.round((absoluteScore / maxScore) * 100, 2));
+        } else {
+            scores.set('relativeScore', 0);
+        }
+        scores.set('maxScore', maxScore);
+        return scores;
+    }
+
+    private round(value, exp) { // helper function to make actually rounding possible
```
Comment (LaKorita) [inline src/main/webapp/app/entities/course/course.service.ts]: Somehow the rounding function doesn't properly work...

Comment (Balgorn) [reply src/main/webapp/app/entities/course/course.service.ts]: can you tell me what is missing my local tests work.

**Comment by LaKorita** [comment]
@Balgorn Please pull the latest changes before making any change. 

Ps. I already fixed what I mentioned in the comments :)

File: src/main/webapp/app/entities/course/course.service.ts (hunk 9 above)
Comment (LaKorita) [reply src/main/webapp/app/entities/course/course.service.ts]: It was a problem related to the calculations that I've mentioned previously in my comments, the method  works. I just changed the number of  decimal point to 1, as I discussed it with Stephan.

Comment (LaKorita) [reply src/main/webapp/app/entities/course/course.service.ts]: Everything is fine now, you don't need to change a thing :)

**Comment by Balgorn** [comment]
@LaKorita already saw that perhaps general comment would be better on my side :)

Anything left I can/should do?
